src/coprocessor/dag/batch_handler.rs

```diff
@@ -0,0 +1,148 @@
+
+            // Check error first, because it means that we should directly respond error.
+            match result.is_drained {
+                Err(Error::Eval(err)) => {
+                    let mut resp = Response::new();
+                    let mut sel_resp = SelectResponse::new();
+                    sel_resp.set_error(err);
+                    let data = box_try!(sel_resp.write_to_bytes());
+                    resp.set_data(data);
+                    return Ok(resp);
+                }
+                Err(e) => return Err(e),
+                Ok(f) => is_drained = f,
+            }
+
+            warnings.merge(&mut result.warnings);
+
+            if result.data.rows_len() > 0 {
+                let mut chunk = Chunk::new();
+                {
+                    let data = chunk.mut_rows_data();
+                    data.reserve(result.data.maximum_encoded_size(&self.output_offsets)?);
+                    result.data.encode(
+                        &self.output_offsets,
+                        &self.executor_context.columns_info,
+                        data,
+                    )?;
+                }
+                chunks.push(chunk);
+            }
+
+            if is_drained {
+                self.out_most_executor
+                    .collect_statistics(&mut self.statistics);
+                self.metrics.cf_stats.add(&self.statistics.cf_stats);
+
+                let mut resp = Response::new();
+                let mut sel_resp = SelectResponse::new();
+                sel_resp.set_chunks(RepeatedField::from_vec(chunks));
+                // FIXME: output_counts should not be i64.
+                sel_resp.set_output_counts(
+                    self.statistics
+                        .scanned_rows_per_range
+                        .iter()
+                        .map(|v| *v as i64)
+                        .collect(),
+                );
+
+                sel_resp.set_warnings(RepeatedField::from_vec(warnings.warnings));
+                sel_resp.set_warning_count(warnings.warning_cnt as i64);
+
+                let data = box_try!(sel_resp.write_to_bytes());
+                resp.set_data(data);
+
+                // Not really useful here, because we only collect it once. But when we change it
+                // in future, hope we will not forget it.
+                self.statistics.clear();
+
+                return Ok(resp);
+            }
+
+            // Grow batch size
+            if batch_size < BATCH_MAX_SIZE {
+                batch_size *= BATCH_GROW_FACTOR;
+                if batch_size > BATCH_MAX_SIZE {
+                    batch_size = BATCH_MAX_SIZE
+                }
+            }
+        }
+    }
+
+    fn collect_metrics_into(&mut self, target_metrics: &mut ExecutorMetrics) {
+        // FIXME: This interface will be broken in streaming mode.
+        target_metrics.merge(&mut self.metrics);
+
+        // Notice: Exec count is collected during building the batch handler.
+    }
+}
```

Comment:
We even don't have unit tests for the non-batch version of DAGHandler for now. Do you have some suggestions about how we can test it (or we should do it later?) @AndreMouche